### ISSUE TITLE
Changing absolute path of mix project causes full recompilation

### ISSUE THREAD

**Issue by jesenko** [elixir-lang/elixir#13241]
### Elixir and Erlang/OTP versions

Elixir 1.15.7 (compiled with Erlang/OTP 26)

### Operating system

Linux

### Current behavior

Create a new mix project `mix new test_me`and compile it using `mix compile`. Then moving to alternate location `mv test_me ./test_me_2` and compile it again with `mix compile`. Whole project is recompiled.

### Expected behavior

In elixir v1.14, project absolute path was ignored when determining if recompilation is needed. The change was intentionally introduced in elixir v1.15 (https://github.com/elixir-lang/elixir/pull/12462), probably to force proper sourcefile metadata to be included in beam modules.

However this change might affect some current deployment / CI caching strategies used in the wild, e.g. caching `_build` and `deps` folder to avoid recompilation.

Not even sure what the expected behaviour should be, as I can see reasoning behind the change (without recompilation, e.g. stacktraces might point to nonexisting file paths?). Maybe additional env variable, which would control if absolute path of project root is considered in recompilation, e.g. `MIX_IGNORE_SOURCE_ROOT=1`.

**Comment by jesenko** [comment]
This form post might also be related to this change: https://elixirforum.com/t/heroku-app-recompiles-after-restarts-using-elixir-1-15

**Comment by sabiwara** [comment]
It might be due to this change: https://github.com/elixir-lang/elixir/pull/12462, which if I remember correctly was needed because the BEAM files themselves are aware of the absolute path.

**Comment by josevalim** [comment]
This is very trick because even `__DIR__` is a full path. So any code using `__DIR__` will also be outdated depending on how `__DIR__` is used. So even if you try to reuse it on CI, you may run into subtle, hard -to-debug bugs.

For the Heroku case, there is a `mix run --no-compile` option or `mix phx.server --no-compile` they could use?

**Comment by jesenko** [comment]
I totally agree with the change in v1.15 regarding the default behaviour. For reuse on CI, probably build path could be made same for all subsequent builds, thus allowing caching of `_build` and `deps`.

For the deployment case (this is where we stumbled onto this) — `--no-compile` flags would be a viable option, but (correct me if I am wrong) `--no-compile` option is available only on mix tasks that explicitly support it. If one however tries to run e.g. `mix ecto.migrate` during deployment procedure, full recompilation is triggered, and no easy way around.

Would it be possible to turn off compilation as environment variable, so that `mix` would avoid compilation altogether for arbitrary task? E.g.:

```
MIX_NO_COMPILE=1 mix ecto.migrate
```

**Comment by josevalim** [comment]
Good point. Either a project configuration or an env var. The project configuration could be set to an env var though. WDYT?

**Comment by jesenko** [comment]
I think that would be great! Probably best if env var could be set even if flag was not explicitly set in project configuration, so that it could be controlled even without changing project source code.

I believe quite some issues could be solved using this env var, e.g. https://elixirforum.com/t/why-does-running-mix-ecto-create-trigger-a-new-compilation/24842

**Comment by josevalim** [comment]
What are your thoughts @sabiwara? Shall we add an env var? We should probably backport to v1.16 too.

**Comment by sabiwara** [comment]
@josevalim do you mean something similar to [`MIX_INSTALL_FORCE`](https://github.com/elixir-lang/elixir/blob/main/lib/mix/lib/mix.ex#L666-L668), and consider either if the [value is `1` or `true`](https://github.com/elixir-lang/elixir/blob/main/lib/mix/lib/mix.ex#L841) or if ?

**Comment by josevalim** [comment]
Yeah, something like `MIX_NO_COMPILE`.

I also looked into calling `mix do compile --no-compile + run` but when the `--no-compile` flag is given, it is not sticky. The next time `compile`is called, it would work. It would work if we allowed some sort of `--never-compile`.

**Comment by sabiwara** [comment]
Will give it a shot 👍

**Comment by josevalim** [comment]
@sabiwara I am thinking maybe the flag is better. It is a bit less obvious but more consistent with everything else in Mix. I am unsure. Thoughts?

**Comment by sabiwara** [comment]
@josevalim sorry I'm taking some time to respond, I'm catching up with the mix codebase.

What is the reason `--no-compile` only works with some tasks?
Would it be possible to actually make it run always? (make it the `--never-compile` flag), or do we need both for composite `aliases`?

**Comment by sabiwara** [comment]
@josevalim do we have a precedent in terms of a "global" flag?

If I understand correctly, it would need to be supported no matter the task, and so we would need to remove it from `args` before passing it to the `c:Mix.Task.run/1`, otherwise it wouldn't work with tasks validating with `Options.Parser`? 

```elixir
➜  goo mx format --never-compile
** (Mix) Could not invoke task "format": 1 error found!
--never-compile : Unknown option
```

The env variable might be easier and might also be more convenient in the context above (CI), since you would need to pass it to every subsequent command: `mix dialyzer --never-compile`, ....

**Comment by josevalim** [comment]
Apologies, I was not clear enough. I do not intend to write a global flag, but rather, add a single flag to `mix compile`. Then you can always invoke a custom compilation by using `mix do`. For example:

    mix do compile --never-compile + phx.server
    mix do compile --never-compile + ecto.migrate

This would have worked with `mix compile --no-compile` except, by design, we allow the "compile" task to be reinvoked if `--no-compile` is given. That's because task A may be fine with `--no-compile` but task B may not. That's why we need a `--no-compile-and-i-make-it-stick` sort of flag.

Does this make sense?

**Comment by jesenko** [comment]
Personally I would prefer env var as it would be more easily set, not needing to refactor all `mix something` calls to `mix do compile --never-compile, something` calls.

As a sidenote - options validation is up-to individual mix task, you can run e.g., `mix ecto.migrate --no-compile`, and the option is just ignored and task proceeds without errors, which initially lead me to believe that `--no-compile` is indeed global option.

**Comment by sabiwara** [comment]
Sorry for the late answer @josevalim.

I understand the `do compile --never-compile +` proposal, but I must confess it feels quite weird to have to call it just not to compile what follows. And it will need to be done for every single task.

The environment variable sounds more practical, although I can see how people could set it while troubleshooting such an issue, forget about it and then find themselves facing some unexpected behavior.

I tried to take a step back and came with an alternative proposal: 
What if we instead offer an `MIX_DIR_UNAWARE_CACHE` env variable (exact naming TBD), which would simply replace `File.cwd!()` by `nil` in the cache key, effectively reverting to the behavior between https://github.com/elixir-lang/elixir/pull/12462.
We could document it with its intended use case and a proper disclaimer, but the risk seems much smaller than simply skipping compilation.

Thoughts?

**Comment by josevalim** [comment]
It depends on where you think the problem is:

1. are we compiling the project wrongly?

2. or are we trying to compile it when the expectation is that you just want to run it?

I believe we must consider `File.cwd!()` in cache because failing to do so leads to wrong builds. So to me it is definitely the second.

However, solving it with `MIX_DIR_UNAWARE_CACHE` or similar is more flexible because we could do it in a way where we allow the build to rellocate, without forcing the user that runs the build to set an env var (i.e. if MIX_DIR_UNAWARE_CACHE was set either during compiling or during checking, we ignore it).

I guess `MIX_NO_COMPILE` could probably find other use cases too. So it is tough.

**Comment by sabiwara** [comment]
Actually just noticed something interesting:
- ecto mix tasks like [`ecto.migrate`](https://hexdocs.pm/ecto_sql/Mix.Tasks.Ecto.Migrate.html#module-command-line-options) **do** offer a `--no-compile` flag
- but it doesn't work reliably in our relocation case
- `mix ecto.migrate --no-compile` is actually flaky: sometimes, it would trigger a full compilation including deps, but sometimes it doesn't

```sh
$ ./relocate.sh

20:40:29.462 [info] Migrations already up
$ ./relocate.sh

20:40:33.439 [info] Migrations already up
$ ./relocate.sh
==> mime
Compiling 1 file (.ex)
Generated mime app
===> Analyzing applications...
===> Compiling telemetry
===> Analyzing applications...
===> Compiling telemetry_poller
==> jason
Compiling 10 files (.ex)
Generated jason app
```

with my `relocate.sh` script being

```sh
set -eu

rm -rf ../copy
mix compile
cp -R . ../copy
cd ../copy

# recompiles!
mix ecto.migrate --no-compile
```

**Comment by josevalim** [comment]
You can try running with `MIX_DEBUG=1`. But it may be we are forgetting to pass the `--no-compile` flag from ecto.migrate to somewhere.

**Comment by sabiwara** [comment]
Actually this doesn't seem to be an ecto only issue, I tried with a custom `Mix.Tasks.Noop` task simply returning `:ok` and I get

```sh
-> Running mix loadconfig (inside Blog.MixProject)
<- Ran mix loadconfig in 2ms
-> Running mix deps.loadpaths (inside Blog.MixProject)
-> Running mix archive.check (inside Blog.MixProject)
<- Ran mix archive.check in 0ms
<- Ran mix deps.loadpaths in 50ms
-> Running mix compile (inside Blog.MixProject)
-> Running mix loadpaths --no-deps-loading (inside Blog.MixProject)
<- Ran mix loadpaths in 0ms
-> Running mix compile.all (inside Blog.MixProject)
-> Running mix compile.yecc (inside Blog.MixProject)
<- Ran mix compile.yecc in 1ms
-> Running mix compile.leex (inside Blog.MixProject)
<- Ran mix compile.leex in 0ms
-> Running mix compile.erlang (inside Blog.MixProject)
<- Ran mix compile.erlang in 0ms
-> Running mix compile.elixir (inside Blog.MixProject)
Compiling 7 files (.ex)
```

We run `deps.loadpaths` before letting the task a chance to set any flag AFAICT.

**Comment by josevalim** [comment]
Is it because the project was not compiled yet? If no task is not found, we always attempt to compile the current project and deps. In any case, I guess your point is that having this flag around `MIX_NO_COMPILE` is going to be brittle, so we should probably make it around `cwd`.

**Comment by sabiwara** [comment]
> Is it because the project was not compiled yet? 

It was compiled just before relocation. Without relocating, it seems we still run `deps.loadpaths` (but it doesn't recompile anything oc):

```sh
-> Running mix deps.loadpaths (inside Blog.MixProject)
-> Running mix archive.check (inside Blog.MixProject)
<- Ran mix archive.check in 0ms
<- Ran mix deps.loadpaths in 64ms
```

but not:

```sh
-> Running mix compile (inside Blog.MixProject)
```

If no task is not found, we always attempt to compile the current project and deps. In any case, I guess your point is that having this flag around MIX_NO_COMPILE is going to be brittle, so we should probably make it around cwd.

For comparison, built-in mix tasks like `MIX_DEBUG=1 mix run --no-compile` would need neither of these and run immediately after `loadconfig`:

```sh
-> Running mix loadconfig (inside Blog.MixProject)
<- Ran mix loadconfig in 3ms
-> Running mix run --no-compile (inside Blog.MixProject)
```

which lets me think we might be compiling the tasks themselves or something?

In summary, before even running the actual task, the following pre-tasks are being run (after `loadconfig` which is always run):

| task \ pre-tasks                      | deps.loadpaths  | compile |
|---------------------------------------|-----------------|---------|
| run --no-compile                      | -               | -       |
| ecto.migrate --no-compile             | called but noop | -       |
| ecto.migrate --no-compile (relocated) | called          | called  |

**Comment by josevalim** [comment]
Yeah, :+1: then with moving forward with a flag called `MIX_COMPILE_CWD_CACHE` which, when disabled, does not check the CWD?

**Comment by sabiwara** [comment]
> Yeah, 👍 then with moving forward with a flag called `MIX_COMPILE_CWD_CACHE` which, when disabled, does not check the CWD?

Will give it a shot!

**Comment by sabiwara** [comment]
Sorry for the slow updates, after checking my attempts against my repro it wasn't solving it.
I've been trying to narrow down the actual root cause, I think I'm getting closer.

Please see this [repo](https://github.com/sabiwara/mix_relocate_repro) with the repro example and some explanations of my findings so far @josevalim.
There seems to be some kind of race condition happening when we touch files.

**Comment by josevalim** [comment]
Thank you for the repo @sabiwara! Mix does everything based on timestamps, so I think your reproduction makes sense? A dependency will be detected as stale based on timestamps too, so preserving modification times is a must.

I guess the issue though is that we are getting stuck because we consider the dependency as stale? We could change the logic to not set the dependency as `:compile` if `MIX_NO_COMPILE` is set but, on the other hand, could we be hiding other issues from not preserving modification times?

**Comment by jannikbecher** [comment]
Maybe this is because of [this](https://apenwarr.ca/log/20181113)?
Andrew Kelley (creator of Zig) mentioned this in a [talk](https://youtu.be/wFlyUzUVFhw?feature=shared&t=1495) he gave and described how he [solved](https://youtu.be/wFlyUzUVFhw?feature=shared&t=2056) it.

**Comment by josevalim** [comment]
It is the opposite problem. We are detecting that files changed but the user does not want to be detected. We already have some “mtime protections” in place, and we could add more to make recompilation on move even more frequent, but that’s the opposite of what is desired here. :) but the issue does confirm going explicit is the best route.

**Comment by sabiwara** [comment]
> Maybe this is because of [this](https://apenwarr.ca/log/20181113)?
> Andrew Kelley (creator of Zig) mentioned this in a [talk](https://youtu.be/wFlyUzUVFhw?feature=shared&t=1495) he gave and described how he [solved](https://youtu.be/wFlyUzUVFhw?feature=shared&t=2056) it.

Thanks @jannikbecher for the links, very interesting! Cache is always a can of worms...

> I guess the issue though is that we are getting stuck because we consider the dependency as stale? We could change the logic to not set the dependency as :compile if MIX_NO_COMPILE is set but, on the other hand, could we be hiding other issues from not preserving modification times?

It makes sense, I can try, but my knowledge of the mix codebase is still insufficient to assess potential undesired side effects.

I thought about this, and I think the most undesirable behavior would be to set this variable, forget about it and then get weird errors without connecting them to the env variable. Then we'd get bug reports that might be hard to reproduce.
Some ideas I had to mitigate it:
- document this flag as experimental in the doc, and see how it goes. This way, we can remove it later if it turns out to be too much of a footgun
- name it in a scary way, like `MIX_DANGEROUSLY_NO_COMPILE`, so people do not use it too casually
- always log something when calling mix with this flag, e.g. `MIX_NO_COMPILE is set, skipping compilation ⚠️`
